src/shared/mkfs-util.c

```diff
@@ -31,16 +32,39 @@ int mkfs_exists(const char *fstype) {
         return true;
 }
 
+static int mangle_fat_label(const char *s, char **ret) {
+        assert(s);
+
+        _cleanup_free_ char *q = NULL;
+        int r;
+
+        r = utf8_to_ascii(s, '_', &q);
+        if (r < 0)
+                return r;
+
+        /* Classic FAT only allows 11 character uppercase labels */
+        q[MIN(strlen(q), 11u)] = '\0';
+        ascii_strupper(q);
+
+        /* mkfs.vfat: Labels with characters *?.,;:/\|+=<>[]" are not allowed.
+         * Let's also replace any control chars. */
+        for (char *p = q; *p; p++)
+                if (strchr("*?.,;:/\\|+=<>[]\"", *p) || char_is_cc(*p))
```

Comment:
maybe say:

```c
if (!strchr(UPPERCASE_LETTERS DIGITS "$%-_@~`!(){}^#"))
        *p = '_';
```

Reply:
I think it's easier to go by what the man page says… E.g. this list misses ' ', '&' and "'".

Reply:
i went with that the FAT spec listed as allowed chars.

point i was making: allowlist is better than denylist here.